File: src/forest.rs
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

use std::fmt;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use ignore::WalkBuilder;
use tree_sitter::{Parser, Tree};

use crate::adapters::{self, LanguageAdapter};
use crate::js_engine;
use crate::rewrite;
use crate::transform;

/// A single parsed source file.
pub struct ParsedFile {
    pub path: PathBuf,
    pub original_source: Vec<u8>,
    pub tree: Tree,
    pub adapter: &'static dyn LanguageAdapter,
}

impl Clone for ParsedFile {
    fn clone(&self) -> Self {
        ParsedFile {
            path: self.path.clone(),
            original_source: self.original_source.clone(),
            tree: self.tree.clone(),
            adapter: self.adapter,
        }
    }
}

/// A pending file change (original + new content).
pub struct FileChange {
    pub path: PathBuf,
    pub original: Vec<u8>,
    pub new_source: Vec<u8>,
}

impl FileChange {
    pub fn diff(&self) -> String {
        rewrite::unified_diff(&self.path, &self.original, &self.new_source)
    }

    pub fn apply(&self) -> Result<()> {
        std::fs::write(&self.path, &self.new_source)
            .with_context(|| format!("writing {}", self.path.display()))
    }
}

/// Apply a set of file changes atomically with backup.
///
/// Strategy:
/// 1. Write all new content to `.canopy.new` temp files
/// 2. Back up all originals to `.canopy.bak`
/// 3. Rename `.new` files to their final paths
///
/// If anything fails during step 3, the `.bak` files allow recovery.
/// Returns the list of backup paths for undo.
pub fn apply_with_backup(changes: &[FileChange]) -> Result<Vec<PathBuf>> {
    let mut temp_paths = Vec::new();
    let mut backup_paths = Vec::new();

    // Step 1: Write new content to temp files.
    for change in changes {
        let temp = change.path.with_extension("canopy.new");
        std::fs::write(&temp, &change.new_source)
            .with_context(|| format!("writing temp {}", temp.display()))?;
        temp_paths.push(temp);
    }

    // Step 2: Back up originals.
    for change in changes {
        let backup = change.path.with_extension("canopy.bak");
        if change.path.exists() {
            std::fs::copy(&change.path, &backup)
                .with_context(|| format!("backing up {}", change.path.display()))?;
        } else {
            // New file — write an empty marker so undo knows to delete it.
            std::fs::write(&backup, b"")
                .with_context(|| format!("creating backup marker {}", backup.display()))?;
        }
        backup_paths.push(backup);
    }

    // Step 3: Rename temp files to final paths.
    for (i, change) in changes.iter().enumerate() {
        // Ensure parent directory exists (for new files).
        if let Some(parent) = change.path.parent() {
            std::fs::create_dir_all(parent)
                .with_context(|| format!("creating directory {}", parent.display()))?;
        }
        std::fs::rename(&temp_paths[i], &change.path)
            .with_context(|| format!("renaming temp to {}", change.path.display()))?;
    }

    Ok(backup_paths)
}

/// Undo a previously applied change by restoring from `.canopy.bak` files.
pub fn undo_from_backups(backup_paths: &[PathBuf]) -> Result<usize> {
    let mut restored = 0;
    for backup in backup_paths {
        // Derive the original path from the backup path.
        let _original = backup.with_extension(""); // strips .canopy.bak
        // Actually, with_extension only strips the last extension.
        // .canopy.bak → .canopy → need to strip again.
        let original = PathBuf::from(backup.to_string_lossy().trim_end_matches(".canopy.bak"));

        if !backup.exists() {
            continue;
        }

        let backup_content = std::fs::read(backup)
            .with_context(|| format!("reading backup {}", backup.display()))?;

        if backup_content.is_empty() && !original.exists() {
            // Empty marker for a file that was newly created — nothing to restore.
            let _ = std::fs::remove_file(backup);
            continue;
        }

        if backup_content.is_empty() {
            // File was newly created — remove it.
            let _ = std::fs::remove_file(&original);
        } else {
            // Restore original content.
            std::fs::write(&original, &backup_content)
                .with_context(|| format!("restoring {}", original.display()))?;
        }

        let _ = std::fs::remove_file(backup);
        restored += 1;
    }
    Ok(restored)
}

/// Clean up backup files after a successful apply (user confirmed the changes are good).
pub fn cleanup_backups(backup_paths: &[PathBuf]) {
    for backup in backup_paths {
        let _ = std::fs::remove_file(backup);
    }
    // Also clean up any stale .new files.
    for backup in backup_paths {
        let new_path = PathBuf::from(
            backup
                .to_string_lossy()
                .replace(".canopy.bak", ".canopy.new"),
        );
        let _ = std::fs::remove_file(new_path);
    }
}

/// A collection of parsed source files.
pub struct Forest {
    pub files: Vec<ParsedFile>,
}

impl Clone for Forest {
    fn clone(&self) -> Self {
        Forest {
            files: self.files.clone(),
        }
    }
}

impl Forest {
    /// Parse all supported files under `path` (file or directory).
    pub fn from_path(path: &Path) -> Result<Self> {
        let mut files = Vec::new();

        if path.is_file() {
            if let Some(parsed) = Self::parse_file(path)? {
                files.push(parsed);
            }
        } else {
            for entry in WalkBuilder::new(path).build() {
                let entry = entry?;
                if entry.file_type().is_some_and(|ft| ft.is_file())
                    && let Some(parsed) = Self::parse_file(entry.path())?
                {
                    files.push(parsed);
                }
            }
        }

        Ok(Forest { files })
    }

    fn parse_file(path: &Path) -> Result<Option<ParsedFile>> {
        let ext = match path.extension().and_then(|e| e.to_str()) {
            Some(ext) => ext,
            None => return Ok(None),
        };

        let adapter = match adapters::adapter_for_extension(ext) {
            Some(a) => a,
            None => return Ok(None),
        };

        let source = std::fs::read(path).with_context(|| format!("reading {}", path.display()))?;

        let mut parser = Parser::new();
        parser
            .set_language(&adapter.language())
            .with_context(|| format!("setting language for {}", path.display()))?;

        let tree = parser
            .parse(&source, None)
            .with_context(|| format!("parsing {}", path.display()))?;

        Ok(Some(ParsedFile {
            path: path.to_owned(),
            original_source: source,
            tree,
            adapter,
        }))
    }

    /// Rename all occurrences of `from` to `to` across the forest.
    /// If `format` is true, runs the language formatter on changed files.
    pub fn rename(&self, from: &str, to: &str, format: bool) -> Result<Vec<FileChange>> {
        let mut changes = Vec::new();

        for file in &self.files {
            let mut new_source = rewrite::rename_in_file(file, from, to)?;
            if new_source != file.original_source {
                if format {
                    new_source = rewrite::format_source(&new_source, file.adapter);
                }
                changes.push(FileChange {
                    path: file.path.clone(),
                    original: file.original_source.clone(),
                    new_source,
                });
            }
        }

        Ok(changes)
    }

    /// Convenience: rename and return unified diff string.
    pub fn rename_diff(&self, from: &str, to: &str) -> Result<String> {
        let changes = self.rename(from, to, false)?;
        Ok(changes.iter().map(|c| c.diff()).collect())
    }

    /// Apply a match/act transform across the forest.
    /// If `format` is true, runs the language formatter on changed files.
    pub fn transform(
        &self,
        match_spec: &transform::Match,
        action: &transform::Action,
        format: bool,
    ) -> Result<Vec<FileChange>> {
        let mut changes = Vec::new();

        for file in &self.files {
            let mut new_source = transform::transform_file(file, match_spec, action)?;
            if new_source != file.original_source {
                if format {
                    new_source = rewrite::format_source(&new_source, file.adapter);
                }
                changes.push(FileChange {
                    path: file.path.clone(),
                    original: file.original_source.clone(),
                    new_source,
                });
            }
        }

        Ok(changes)
    }

    /// Apply a JS transform function across the forest.
    pub fn transform_js(
        &self,
        match_spec: &transform::Match,
        transform_fn: &str,
        format: bool,
    ) -> Result<Vec<FileChange>> {
        let mut changes = Vec::new();

        for file in &self.files {
            let query_str = transform::resolve_query_str(file.adapter, match_spec)?;
            let mut new_source = js_engine::run_js_transform(
                &file.original_source,
                &file.tree,
                &query_str,
                transform_fn,
                &file.path.to_string_lossy(),
                file.adapter,
            )?;
            if new_source != file.original_source {
                if format {
                    new_source = rewrite::format_source(&new_source, file.adapter);
                }
                changes.push(FileChange {
                    path: file.path.clone(),
                    original: file.original_source.clone(),
                    new_source,
                });
            }
        }

        Ok(changes)
    }

    /// Query the forest for nodes matching a pattern.
    /// Returns a list of (file_path, line, column, node_kind, node_text) tuples.
    pub fn query(&self, match_spec: &transform::Match) -> Result<Vec<QueryResult>> {
        let mut results = Vec::new();

        for file in &self.files {
            let matches = transform::query_file(file, match_spec)?;
            results.extend(matches);
        }

        Ok(results)
    }
}

/// A single query match result.
pub struct QueryResult {
    pub path: PathBuf,
    pub start_line: usize,
    pub start_col: usize,
    pub kind: String,
    pub name: Option<String>,
    pub text: String,
}

impl fmt::Display for Forest {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Forest: {} file(s)", self.files.len())?;
        for file in &self.files {
            let has_errors = file.tree.root_node().has_error();
            let status = if has_errors { " [parse errors]" } else { "" };
            writeln!(f, "  {}{status}", file.path.display())?;
        }
        Ok(())
    }
}

File: go/forest/forest.go
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

// Package forest provides types and functions for loading, parsing, and
// manipulating collections of source files via Tree-sitter.
package forest

import (
	"bytes"
	"errors"
	"fmt"
	"io/fs"

	"github.com/marcelocantos/sawmill/paths"
	"os"
	"path/filepath"
	"strings"

	tree_sitter "github.com/marcelocantos/sawmill/tscompat"

	"github.com/marcelocantos/sawmill/adapters"
)

// ParsedFile represents a single parsed source file.
type ParsedFile struct {
	Path           string
	OriginalSource []byte
	Tree           *tree_sitter.Tree
	Adapter        adapters.LanguageAdapter
}

// FileChange represents a pending file change (original + new content).
type FileChange struct {
	Path      string
	Original  []byte
	NewSource []byte
}

// Diff returns a unified diff between the original and new content.
// A diffFn is accepted to avoid a circular import between forest and rewrite.
// Callers typically pass rewrite.UnifiedDiff.
func (fc *FileChange) Diff(diffFn func(path string, original, newContent []byte) string) string {
	return diffFn(fc.Path, fc.Original, fc.NewSource)
}

// Apply writes the new content directly to the file's path.
func (fc *FileChange) Apply() error {
	if err := os.WriteFile(fc.Path, fc.NewSource, 0o644); err != nil {
		return fmt.Errorf("writing %s: %w", fc.Path, err)
	}
	return nil
}

// Forest is a collection of parsed source files.
type Forest struct {
	Files []*ParsedFile
}

// SkipFn decides whether to skip a directory during a walk. Returns true to
// skip. A nil SkipFn skips nothing.
type SkipFn func(absPath string) bool

// FromPath parses all recognised source files under path (file or directory).
// All directories are walked. Callers that want to skip directories should
// use FromPathSkip with a SkipFn (e.g. backed by scope.Classifier.ShouldSkipDir).
func FromPath(path string) (*Forest, error) {
	return FromPathSkip(path, nil)
}

// FromPathSkip is FromPath with a directory-skip predicate.
func FromPathSkip(path string, skip SkipFn) (*Forest, error) {
	var files []*ParsedFile

	info, err := os.Stat(path)
	if err != nil {
		return nil, fmt.Errorf("stat %s: %w", path, err)
	}

	if !info.IsDir() {
		parsed, err := parseFile(path)
		if err != nil {
			return nil, err
		}
		if parsed != nil {
			files = append(files, parsed)
		}
		return &Forest{Files: files}, nil
	}

	err = filepath.WalkDir(path, func(p string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if d.IsDir() {
			if p != path && skip != nil && skip(p) {
				return filepath.SkipDir
			}
			return nil
		}
		parsed, err := parseFile(p)
		if err != nil {
			return err
		}
		if parsed != nil {
			files = append(files, parsed)
		}
		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("walking %s: %w", path, err)
	}

	return &Forest{Files: files}, nil
}

// parseFile reads and parses a single file, returning nil if the extension
// is not recognised by any adapter.
func parseFile(path string) (*ParsedFile, error) {
	ext := strings.TrimPrefix(filepath.Ext(path), ".")
	if ext == "" {
		return nil, nil
	}

	adapter := adapters.ForExtension(ext)
	if adapter == nil {
		return nil, nil
	}

	source, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading %s: %w", path, err)
	}

	parser := tree_sitter.NewParser()
	defer parser.Close()

	if err := parser.SetLanguage(adapter.Language()); err != nil {
		return nil, fmt.Errorf("setting language for %s: %w", path, err)
	}

	tree := parser.Parse(source, nil)
	if tree == nil {
		return nil, fmt.Errorf("parsing %s: tree-sitter returned nil tree", path)
	}

	return &ParsedFile{
		Path:           path,
		OriginalSource: source,
		Tree:           tree,
		Adapter:        adapter,
	}, nil
}

// String returns a human-readable summary of the forest.
func (f *Forest) String() string {
	var sb strings.Builder
	fmt.Fprintf(&sb, "Forest: %d file(s)\n", len(f.Files))
	for _, file := range f.Files {
		status := ""
		if file.Tree.RootNode().HasError() {
			status = " [parse errors]"
		}
		fmt.Fprintf(&sb, "  %s%s\n", file.Path, status)
	}
	return sb.String()
}

// QueryResult is a single query match result.
type QueryResult struct {
	Path      string
	StartLine uint
	StartCol  uint
	Kind      string
	Name      string // empty if not applicable
	Text      string
}

// ErrStaleContent is returned by ApplyWithBackup when a target file's on-disk
// content no longer matches the FileChange.Original it was previewed against —
// i.e. the file changed between preview and apply (a concurrent apply, an
// external edit, or a watcher-observed reparse). Applying anyway would silently
// clobber the newer content (a lost update), so the whole apply is aborted.
var ErrStaleContent = errors.New("file changed on disk since preview")

// BackupEntry records everything needed to reverse one file's change on undo.
type BackupEntry struct {
	// OriginalPath is the absolute path of the changed file.
	OriginalPath string
	// BackupFile is the absolute path of the saved pre-apply copy. Empty when
	// Existed is false (the file was created by this apply and undo removes it).
	BackupFile string
	// Existed reports whether OriginalPath existed before the apply. Tracked
	// explicitly rather than inferred from empty backup content, so an
	// originally-empty file is restored to empty on undo instead of deleted.
	Existed bool
	// Mode is the original file mode, re-applied on both apply and undo so a
	// transform never silently strips (e.g.) the execute bit. Meaningful only
	// when Existed is true.
	Mode fs.FileMode
}

// ApplyManifest is the undo record produced by ApplyWithBackup: the per-apply
// staging/backup directory plus one entry per changed file.
type ApplyManifest struct {
	Dir     string
	Entries []BackupEntry
}

// ApplyWithBackup applies a set of file changes atomically with backups stored
// under a fresh per-apply directory inside ~/.sawmill/backups/ (never in the
// project tree).
//
// Strategy:
//  0. Pre-flight: verify each existing target still matches FileChange.Original
//     (optimistic-concurrency guard); abort the whole apply on mismatch before
//     writing anything.
//  1. Write all new content to staging files, carrying the original file mode.
//  2. Back up all originals, recording existence + mode in the manifest.
//  3. Rename staging files into place; on a partial failure, roll every
//     already-renamed file back to its pre-apply state.
//
// Returns the manifest (consumed by UndoFromBackups) on success.
func ApplyWithBackup(root string, changes []FileChange) (*ApplyManifest, error) {
	// Step 0: Optimistic-concurrency pre-flight. Read-only; nothing is written
	// until every target is confirmed to still match what the preview saw.
	for _, change := range changes {
		info, err := os.Stat(change.Path)
		if err != nil || info.IsDir() {
			continue // new file (or a directory we can't back up) — no baseline
		}
		current, rerr := os.ReadFile(change.Path)
		if rerr != nil {
			return nil, fmt.Errorf("reading %s for concurrency check: %w", change.Path, rerr)
		}
		if !bytes.Equal(current, change.Original) {
			return nil, fmt.Errorf("%s: %w", change.Path, ErrStaleContent)
		}
	}

	dir, err := paths.NewApplyDir(root)
	if err != nil {
		return nil, fmt.Errorf("creating apply dir: %w", err)
	}

	manifest := &ApplyManifest{Dir: dir}
	tempPaths := make([]string, len(changes))

	// Step 1: Write new content to staging files, preserving the original mode
	// so the post-rename inode keeps its permission bits.
	for i, change := range changes {
		mode := fs.FileMode(0o644)
		if info, err := os.Stat(change.Path); err == nil {
			mode = info.Mode().Perm()
		}
		temp := filepath.Join(dir, fmt.Sprintf("%d.new", i))
		if err := os.WriteFile(temp, change.NewSource, mode); err != nil {
			return nil, fmt.Errorf("writing temp %s: %w", temp, err)
		}
		tempPaths[i] = temp
	}

	// Step 2: Back up originals with explicit existence + mode.
	manifest.Entries = make([]BackupEntry, len(changes))
	for i, change := range changes {
		entry := BackupEntry{OriginalPath: change.Path}
		if info, err := os.Stat(change.Path); err == nil {
			entry.Existed = true
			entry.Mode = info.Mode().Perm()
			entry.BackupFile = filepath.Join(dir, fmt.Sprintf("%d.bak", i))
			if err := copyFile(change.Path, entry.BackupFile); err != nil {
				return nil, fmt.Errorf("backing up %s: %w", change.Path, err)
			}
		}
		manifest.Entries[i] = entry
	}

	// Step 3: Rename staging files into place. On any failure, restore every
	// file already renamed this pass so the tree is left in its pre-apply state.
	var renamed []int
	for i, change := range changes {
		if parent := filepath.Dir(change.Path); parent != "" {
			if err := os.MkdirAll(parent, 0o755); err != nil {
				rollbackRenames(manifest, renamed)
				return nil, fmt.Errorf("creating directory %s: %w", parent, err)
			}
		}
		if err := os.Rename(tempPaths[i], change.Path); err != nil {
			rollbackRenames(manifest, renamed)
			return nil, fmt.Errorf("renaming temp to %s: %w", change.Path, err)
		}
		// os.Rename replaces the inode with the staging temp's; re-assert the
		// original mode exactly so umask can't alter it.
		if manifest.Entries[i].Existed {
			_ = os.Chmod(change.Path, manifest.Entries[i].Mode)
		}
		renamed = append(renamed, i)
	}

	return manifest, nil
}

// rollbackRenames reverses the already-completed renames of a failed apply,
// restoring each file from its backup (or removing it if it was newly created),
// then discards the per-apply directory. Best effort: rolling back a
// half-applied change must not itself abort partway and leave things worse.
func rollbackRenames(manifest *ApplyManifest, renamed []int) {
	for j := len(renamed) - 1; j >= 0; j-- {
		e := manifest.Entries[renamed[j]]
		if !e.Existed {
			_ = os.Remove(e.OriginalPath)
			continue
		}
		if data, err := os.ReadFile(e.BackupFile); err == nil {
			_ = os.WriteFile(e.OriginalPath, data, e.Mode)
			_ = os.Chmod(e.OriginalPath, e.Mode)
		}
	}
	_ = os.RemoveAll(manifest.Dir)
}

// UndoFromBackups restores every file recorded in the manifest to its
// pre-apply state and removes the per-apply directory. Files that did not
// exist before the apply are deleted; files that existed (including
// originally-empty ones) are rewritten with their saved content and mode.
// Returns the number of files successfully restored.
func UndoFromBackups(manifest *ApplyManifest) (int, error) {
	if manifest == nil {
		return 0, nil
	}
	restored := 0
	for i := len(manifest.Entries) - 1; i >= 0; i-- {
		e := manifest.Entries[i]
		if !e.Existed {
			// File was created by the apply — remove it.
			if err := os.Remove(e.OriginalPath); err != nil && !os.IsNotExist(err) {
				return restored, fmt.Errorf("removing %s: %w", e.OriginalPath, err)
			}
			restored++
			continue
		}
		data, err := os.ReadFile(e.BackupFile)
		if err != nil {
			return restored, fmt.Errorf("reading backup %s: %w", e.BackupFile, err)
		}
		if err := os.WriteFile(e.OriginalPath, data, e.Mode); err != nil {
			return restored, fmt.Errorf("restoring %s: %w", e.OriginalPath, err)
		}
		if err := os.Chmod(e.OriginalPath, e.Mode); err != nil {
			return restored, fmt.Errorf("chmod %s: %w", e.OriginalPath, err)
		}
		restored++
	}
	_ = os.RemoveAll(manifest.Dir)
	return restored, nil
}

// CleanupBackups removes a completed apply's staging/backup directory, called
// when the backups are no longer needed for undo.
func CleanupBackups(manifest *ApplyManifest) {
	if manifest == nil {
		return
	}
	_ = os.RemoveAll(manifest.Dir)
}

// copyFile copies src to dst, preserving content and permissions.
func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	info, err := os.Stat(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, data, info.Mode())
}
